src/preload/preload.js

```diff
@@ -1,141 +1,386 @@
+ * @param {string} url 
+ * @param {FormData=} formData
+ */
+function sendXmlGetRequest(url, sourceNode, formData = undefined) {
+  const xhr = new XMLHttpRequest()
+  if (formData) {
+    url += '?' + new URLSearchParams(formData.entries()).toString()
   }
-})
+  xhr.open('GET', url);
+  xhr.setRequestHeader("HX-Preloaded", "true")
+  xhr.onload = function() { processResponse(sourceNode, xhr.responseText) }
+  xhr.send()
+}
+
+/**
+ * Process request response by marking node `DONE` to prevent repeated
+ * requests, except if preload attribute contains `always`,
+ * and load linked resources (e.g. images) returned in the response 
+ * if `preload-images` attribute is `true`
+ * @param {Node} node 
+ * @param {string} responseText 
+ */
+function processResponse(node, responseText) {
+  node.preloadState = node.preloadAlways ? 'READY' : 'DONE'
+
+  if (getClosestAttribute(node, 'preload-images') === 'true') {
+    // Load linked resources
+    document.createElement('div').innerHTML = responseText
+  }
+}
+
+/**
+ * Gets attribute value from node or one of its parents
+ * @param {Node} node 
+ * @param {string} attribute 
+ * @returns { string | undefined }
+ */
+function getClosestAttribute(node, attribute) {
+  if (node == undefined) { return undefined }
+  return node.getAttribute(attribute)
+    || node.getAttribute('data-' + attribute)
+    || getClosestAttribute(node.parentElement, attribute)
+}
+
+/**
+ * Determines if node is valid for preloading and should be
+ * initialized by setting up event listeners and handlers
+ * @param {Node} node 
+ * @returns {boolean}
+ */
+function isValidNodeForPreloading(node) {
+  // Add listeners only to nodes which include "GET" transactions
+  // or preloadable "GET" form elements
+  const getReqAttrs = ['href', 'hx-get', 'data-hx-get'];
+  const includesGetRequest = node => getReqAttrs.some(a => node.hasAttribute(a))
+                                      || node.method == 'get';
+  const isPreloadableGetFormElement = node.form
+                                      && includesGetRequest(node.form)
+                                      && isPreloadableFromElement(node)
+  if (!includesGetRequest(node) && !isPreloadableFromElement) {
+    return false
+  }
+
+  // Don't preload <input> elements contained in <label>
+  // to prevent sending two requests
+  if (node.tagName === 'INPUT' && node.closest('label')) {
+    return false
+  }
+
+  return true
+}
+
+/**
+ * Determine if node is a form element which can be preloaded,
+ * i.e., `radio`, `checkbox`, `select` or `submit` button
+ * or a `label` of a form element which can be preloaded.
+ * @param {Node} node 
+ * @returns {boolean}
+ */
+function isPreloadableFromElement(node) {
+  if (node instanceof HTMLInputElement) {
+    const preloadableInputTypes = ['checkbox', 'radio', 'submit'];
+    return preloadableInputTypes.some(
+      type => node.getAttribute('type') === type);
```

Comment:
You are right - this code is too complex. I have simplified it to make sure `getAttribute` is called only once. I have replaced `some` with `includes`. I believe `includes` is more elegant than `indexOf`, which you proposed, because we don't actually use the index anywhere. We just want to check if the type is among the types for which the extension currently supports smart preloading.

Reply:
Yep you're totally right, as we were coding in a IE-11 compatible way for htmx 1 (which doesn't support include), I still have some IE-11 friendly code habits sometimes, sigh... `includes` is perfect here!